extensions/infinispan-client/deployment/src/main/java/io/quarkus/infinispan/client/deployment/InfinispanClientProcessor.java

```diff
@@ -281,4 +463,160 @@ void registerServiceBinding(Capabilities capabilities, BuildProducer<ServiceProv
+                    remoteCacheBean.clientName = clientNameQualifier == null ? DEFAULT_INFINISPAN_CLIENT_NAME
+                            : clientNameQualifier.value().asString();
+                    return remoteCacheBean;
+                })
+                .collect(Collectors.toSet());
+
+        if (!clientNames.contains(DEFAULT_INFINISPAN_CLIENT_NAME)) {
+            boolean createDefaultCacheManager = finishedBuildItem.getInjectionPoints().stream()
+                    .filter(i -> SUPPORTED_INJECTION_TYPE.contains(i.getRequiredType().name())
+                            && i.getRequiredQualifier(INFINISPAN_CLIENT_ANNOTATION) == null)
+                    .findAny()
+                    .isPresent();
+            if (createDefaultCacheManager) {
+                clientNames.add(DEFAULT_INFINISPAN_CLIENT_NAME);
+            }
+        }
+
+        // Produce default and/or named RemoteCacheManager and CounterManager beans
+        for (String clientName : clientNames) {
+            syntheticBeanBuildItemBuildProducer.produce(
+                    configureAndCreateSyntheticBean(clientName, RemoteCacheManager.class,
+                            recorder.infinispanClientSupplier(clientName)));
+            syntheticBeanBuildItemBuildProducer.produce(
+                    configureAndCreateSyntheticBean(clientName, CounterManager.class,
+                            recorder.infinispanConterManagerSupplier(clientName)));
+        }
+
+        // Produce RemoteCache beans
+        for (RemoteCacheBean remoteCacheBean : remoteCacheBeans) {
+            syntheticBeanBuildItemBuildProducer.produce(
+                    configureAndCreateSyntheticBean(remoteCacheBean,
+                            recorder.infinispanRemoteCacheClientSupplier(remoteCacheBean.clientName,
+                                    remoteCacheBean.cacheName)));
+        }
+    }
+
+    static <T> SyntheticBeanBuildItem configureAndCreateSyntheticBean(String name,
+            Class<T> type,
+            Supplier<T> supplier) {
+        SyntheticBeanBuildItem.ExtendedBeanConfigurator configurator = buildSyntheticBean(
+                SyntheticBeanBuildItem.configure(type).supplier(supplier));
+
+        if (InfinispanClientUtil.isDefault(name)) {
+            configurator.addQualifier(Default.class);
+        } else {
+            configurator.addQualifier().annotation(DotNames.NAMED).addValue("value", name).done();
+            configurator.addQualifier().annotation(INFINISPAN_CLIENT_ANNOTATION).addValue("value", name).done();
+        }
+
+        return configurator.done();
+    }
+
+    static <T> SyntheticBeanBuildItem configureAndCreateSyntheticBean(RemoteCacheBean remoteCacheBean, Supplier<T> supplier) {
+        SyntheticBeanBuildItem.ExtendedBeanConfigurator configurator = SyntheticBeanBuildItem.configure(RemoteCache.class)
+                .types(remoteCacheBean.type)
+                .scope(Singleton.class) // Some Infinispan API won't work if this is not a Mock
```

Comment:
What does the comment mean? The bean itself is not a mock and it's not really mockable unless it uses client proxy; i.e. uses `@ApplicationScoped` instead of `@Singleton` in this case.

Reply:
Yes, it's a limitation we have with this version of Infinispan. API Search.from(cache) breaks. We need to make evolutions to make this possible, and possibly this will be doable with the new api we are finishing to develop but not with this api

Reply:
Ok, my understanding from our Zulip conversation was that you need these beans to be proxied, so I am surprised to see `Singleton.class` there.

Reply:
yes, I was also disappointed when I saw the search code failing due to a class cast exception... we need changes in the infinispan search code to make this work, but we are considering that might be better to make the effort in the new api that we are finishing for infinispan 15